sdks/rs/src/attestation.rs

```diff
@@ -30,57 +35,72 @@ pub enum AttestationError {
     HttpBodyError(#[from] hyper::Error),
 }
 
-fn get_all_certs(cert: X509, cabundle: Vec<Value>) -> Result<Vec<X509>, AttestationError> {
-    let mut all_certs = vec![cert];
-    for cert in cabundle {
-        let cert = (match cert {
-            Value::Bytes(b) => Ok(b),
-            _ => Err(AttestationError::ParseFailed("cert decode".into())),
-        })?;
-        let cert = X509::from_der(&cert)
-            .map_err(|e| AttestationError::ParseFailed(format!("der: {e}")))?;
-        all_certs.push(cert);
-    }
-    Ok(all_certs)
+#[derive(Debug, Default)]
+pub struct AttestationExpectations {
+    pub timestamp: Option<usize>,
+    // (max age, current timestamp)
+    pub age: Option<(usize, usize)>,
+    pub pcrs: Option<[[u8; 48]; 3]>,
+    pub root_public_key: Option<Vec<u8>>,
 }
 
-fn verify_cert_chain(
-    cert: X509,
-    cabundle: Vec<Value>,
-    root_cert_pem: Vec<u8>,
-) -> Result<(), AttestationError> {
-    let certs = get_all_certs(cert, cabundle)?;
+pub fn verify(
+    attestation_doc: Vec<u8>,
+    expectations: AttestationExpectations,
+) -> Result<AttestationDecoded, AttestationError> {
+    let mut result = AttestationDecoded {
+        pcrs: [[0; 48]; 3],
+        timestamp: 0,
+        root_public_key: Vec::new(),
+        public_key: Vec::new(),
+    };
 
-    for i in 0..(certs.len() - 1) {
-        let pubkey = certs[i + 1]
-            .public_key()
-            .map_err(|e| AttestationError::ParseFailed(format!("pubkey: {e}")))?;
-        if !certs[i]
-            .verify(&pubkey)
-            .map_err(|e| AttestationError::ParseFailed(format!("signature: {e}")))?
-        {
-            return Err(AttestationError::VerifyFailed("signature".into()));
+    // parse attestation doc
+    let (cosesign1, mut attestation_doc) = parse_attestation_doc(&attestation_doc)?;
+
+    // parse timestamp
+    result.timestamp = parse_timestamp(&mut attestation_doc)?;
+
+    // check expected timestamp if exists
+    if let Some(expected_ts) = expectations.timestamp {
+        if result.timestamp != expected_ts {
+            return Err(AttestationError::VerifyFailed("timestamp mismatch".into()));
         }
-        if certs[i + 1].issued(&certs[i]) != X509VerifyResult::OK {
-            return Err(AttestationError::VerifyFailed("issuer or subject".into()));
+    }
+
+    // check age if exists
+    if let Some((max_age, current_ts)) = expectations.age {
+        if result.timestamp <= current_ts && current_ts - result.timestamp > max_age {
```

Comment:
1st condition is redundant. Only second condition is sufficient.

Reply:
it would underflow and panic no?

Reply:
yeah, right.